src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs

```diff
@@ -0,0 +1,100 @@
+            Debug.Assert(cookieContainer != null);
+
+            // Get 'Set-Cookie' headers from response
+            List<string> cookieHeaders =
+                WinHttpResponseParser.GetResponseHeaders(requestHandle, Interop.WinHttp.WINHTTP_QUERY_SET_COOKIE);
+            WinHttpTraceHelper.Trace("WINHTTP_QUERY_SET_COOKIE");
+            
+            foreach (string cookieHeader in cookieHeaders)
+            {
+                WinHttpTraceHelper.Trace(cookieHeader);
+                try
+                {
+                    cookieContainer.SetCookies(request.RequestUri, cookieHeader);
+                }
+                catch (CookieException)
+                {
+                    // We ignore malformed cookies in the response.
+                }
+            }            
+        }
+        
+        public static void ResetCookieRequestHeaders(WinHttpRequestState state, Uri redirectUri)
+        {
+            SafeWinHttpHandle requestHandle = state.RequestHandle;
+            
+            Debug.Assert(state.Handler.CookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer);
+            
+            // Clear cookies.
+            if (!Interop.WinHttp.WinHttpAddRequestHeaders(
+                requestHandle,
+                CookieHeaderNameWithColon,
+                (uint)CookieHeaderNameWithColon.Length,
+                Interop.WinHttp.WINHTTP_ADDREQ_FLAG_REPLACE))
+            {
+                int lastError = Marshal.GetLastWin32Error();
+                if (lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+                {
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+            }
+
+            // Re-add cookies. The GetCookieHeader() method will return the correct set of
+            // cookies based on the redirectUri.
+            string cookieHeader = GetCookieHeader(redirectUri, state.Handler.CookieContainer);
+            if (!string.IsNullOrEmpty(cookieHeader))
+            {
+                if (!Interop.WinHttp.WinHttpAddRequestHeaders(
+                    requestHandle,
+                    cookieHeader,
+                    (uint)cookieHeader.Length,
+                    Interop.WinHttp.WINHTTP_ADDREQ_FLAG_ADD))
+                {
+                    WinHttpException.ThrowExceptionUsingLastError();
+                }
+            }
+        }
+        
+        public static string GetCookieHeader(Uri uri, CookieContainer cookies)
+        {
+            string cookieHeader = null;
+
+            Debug.Assert(cookies != null);
+
+            string cookieValues = cookies.GetCookieHeader(uri);
+            if (!string.IsNullOrEmpty(cookieValues))
+            {
+                cookieHeader = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0} {1}",
+                    CookieHeaderNameWithColon,
+                    cookieValues);
```

Comment:
`cookieHeader = HttpKnownHeaderNames.Cookie + ": " + cookieValues;` should be more efficient than `string.Format` and avoids having to redefine the `Cookie` constant.